### PR TITLE
ENT-3473: Transmitter for blackboard, sends courseId as a hash of course Key

### PR DESCRIPTION
**Pull request by binodpant**
**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [x] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [x] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [x] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [Travis](https://travis-ci.org/github/edx/edx-enterprise), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: integrated_channels/blackboard/client.py
Comment (taliaedX) [inline integrated_channels/blackboard/client.py:48]: Wouldn't this normally be a noop anyways? IE you should never even call this method if you don't have data?

File: integrated_channels/blackboard/exporters/content_metadata.py
Comment (taliaedX) [inline integrated_channels/blackboard/exporters/content_metadata.py:43]: I feel like there's gotta be a simpler way then 3 .replace

File: integrated_channels/blackboard/transmitters/content_metadata.py
Comment (taliaedX) [inline integrated_channels/blackboard/transmitters/content_metadata.py:27]: but what about chunk_size default of 1?

**Review by taliaedX — COMMENTED** [review]
some questions

File: integrated_channels/blackboard/exporters/content_metadata.py
Comment (binodpant) [reply integrated_channels/blackboard/exporters/content_metadata.py]: replaced with client cdoe of hash instead...

File: integrated_channels/blackboard/transmitters/content_metadata.py
Comment (binodpant) [reply integrated_channels/blackboard/transmitters/content_metadata.py:27]: ah yeah need to do that too thanks for the reminder...

File: integrated_channels/blackboard/client.py
Comment (binodpant) [reply integrated_channels/blackboard/client.py:49]: yeah I think we can scrap this one and assume we will get valid data otherwise we will fail later anyway

File: integrated_channels/blackboard/client.py
Comment (alex-sheehan-edx) [inline integrated_channels/blackboard/client.py:68]: A good practice that I like from Talia's work is wrapping the requests/error handling (checking status code). I added helper methods for posts, gets and patches in my PR if you want to use those!

File: integrated_channels/blackboard/client.py
Comment (alex-sheehan-edx) [inline integrated_channels/blackboard/client.py:130]: 👍

File: integrated_channels/blackboard/exporters/content_metadata.py
Comment (alex-sheehan-edx) [inline integrated_channels/blackboard/exporters/content_metadata.py:15]: perhaps an explanation as to why the courseId is random/why it's ok that the course id is random?

File: integrated_channels/integrated_channel/management/commands/__init__.py
Comment (alex-sheehan-edx) [inline integrated_channels/integrated_channel/management/commands/__init__.py:23]: 🎉 alphabetical 🎉

**Review by alex-sheehan-edx — COMMENTED** [review]
Overall, looks pretty good! Just a couple q's and we should definitely work together while merging.

File: integrated_channels/blackboard/exporters/content_metadata.py
Comment (binodpant) [reply integrated_channels/blackboard/exporters/content_metadata.py:15]: yup I think the hash function works as designed, seeds at python subsystem startup time so we can't use it as is if we want consistent hash values ...but I can see this comment being a bit confusing here (why even mention that issue here, since there is no code here to generate courseId)

File: integrated_channels/blackboard/client.py
Comment (binodpant) [reply integrated_channels/blackboard/client.py]: hmm, not clear on how to do that, where are your helpers? I only have helpers here for oauth and auth header creation

Comment (binodpant) [reply integrated_channels/blackboard/client.py]: I think I see yours now https://github.com/edx/edx-enterprise/pull/998/files#diff-e8224b0f5aaaee88283c12d5d533f5700f33f2e63c7939e93f814ef94bf95716R122 yeah I guess we could just have error handling inside the _put/_post methods. Typically I don't consider it a great idea to have automatic error handling in lower level methods since 400 is an issue to be dealt with by whoever calls _put/_post etc while the lower level put just does the putting.

Also if we have error handling in _put/_post we don't have the ability to customize error messages  (unless we catch and rethrow?)

Somehow I will find a way to leverage your code once it merges while also being able to throw exceptions with custom messages

Comment (binodpant) [reply integrated_channels/blackboard/client.py]: also noting thatt your code does return the response_text along with the code so it's actually going to be quite good I thin even as is and I can try not even having any error handling in my code initially then see if I need to tune messages

**Comment by binodpant** [comment]
Sorry for lots of small commits. Let's hold on reviewing this until I handle update and possibly delete

but otherwise other code/tests are in place

will hopefully get to all this by tomorrow

**Comment by alex-sheehan-edx** [comment]
> Sorry for lots of small commits. Let's hold on reviewing this until I handle update and possibly delete
> 
> but otherwise other code/tests are in place
> 
> will hopefully get to all this by tomorrow

What you have now LGTM. Will review quickly if you push the changes for update and such.

File: integrated_channels/blackboard/client.py
Comment (taliaedX) [inline integrated_channels/blackboard/client.py:65]: why perform a copy if you're just appending a key that wasn't in there in the first place? (I assume 'courseId' is otherwise not populated)

Comment (binodpant) [reply integrated_channels/blackboard/client.py:65]: yeah I was just being a good citizen wrt not modifying input vars.. not sure if we utilize this pattern much around our codebase

Comment (taliaedX) [reply integrated_channels/blackboard/client.py:65]: Hmm yeah that's very important in React, but I am not so sure about Python. I'm not 100% up on the best practices though.

I will say that copying variables is definitely not the best practice when dealing with large amounts of data, so it's something to keep in mind. I don't think the data we're dealing with here is particularly large, but I also don't know where this runs precisely and how low of memory limits we may be dealing with.

For instance, in a previous project at a previous company, we were doing a lot of data transformations on csv files, some of which were several gbs in size. Originally because of the different transformations stacked on top of each other, we were copying the original data variable 4+ times. We crashed instances many, many times. These instances had 32+ gb of memory. 😅 

So TL;DR: I don't know if it is normally a best practice or not, but for performance reasons, I'm generally wary of it, but we also probably deal with so little data that it's not a big deal here.

Comment (binodpant) [reply integrated_channels/blackboard/client.py:65]: yeah totally great points, we have to be watchful of performance when doing this for sure. in this case yeah ti seem innocuous enough to try at least, thanks

File: integrated_channels/blackboard/client.py
Comment (binodpant) [inline integrated_channels/blackboard/client.py:124]: renamed a couple vars in this function, fyi, for clarity

File: tests/test_integrated_channels/test_blackboard/test_client.py
Comment (binodpant) [inline tests/test_integrated_channels/test_blackboard/test_client.py:170]: refactored client a bit so we enclose the fetching of blackboard course id into one function, hence this change

File: integrated_channels/blackboard/client.py
Comment (taliaedX) [inline integrated_channels/blackboard/client.py:28]: I don't question the accuracy, this is just a question for my understanding: Why are we using v1-v3 versions? Do they really not have the functionality all in a "v3"?

File: integrated_channels/blackboard/client.py
Comment (taliaedX) [inline integrated_channels/blackboard/client.py:48]: why not use the `COURSE_PATH` var here (and admittedly update it to be v3 instead of v1)?

File: integrated_channels/blackboard/models.py
Comment (taliaedX) [inline integrated_channels/blackboard/models.py:68]: 👍

**Review by taliaedX — APPROVED** [review]
Looking good to me! A question or 2 around API versions we're accessing, but really no issues here.

File: integrated_channels/blackboard/client.py
Comment (binodpant) [reply integrated_channels/blackboard/client.py:28]: ya they are transitioning slowly it seems

File: integrated_channels/blackboard/client.py
Comment (binodpant) [reply integrated_channels/blackboard/client.py:48]: and I have a bug here! should have tested locally, my url is not complete ... will fix and be back

Comment (binodpant) [reply integrated_channels/blackboard/client.py:48]: and I can't use the existing vars due to the way they are setup..

Comment (taliaedX) [reply integrated_channels/blackboard/client.py:48]: so they really don't have their get all courses endpoint on v3, but they do have their get/delete individual course on v3?

File: integrated_channels/blackboard/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/blackboard/client.py:28]: docs are a mix of versions. Seems like certain areas are farther developed than others. Unfortunately, don't think we have much options here.

File: integrated_channels/blackboard/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/blackboard/client.py:48]: why isn't this a constant with the other paths?

File: integrated_channels/blackboard/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/blackboard/client.py:124]: that's fine with me :)

File: integrated_channels/blackboard/client.py
Comment (alex-sheehan-edx) [inline integrated_channels/blackboard/client.py:58]: Don't you have a static method for this?

File: integrated_channels/blackboard/client.py
Comment (alex-sheehan-edx) [inline integrated_channels/blackboard/client.py:190]: if you're searching by externalId on Blackboards side, will this ever not be true?

Comment (alex-sheehan-edx) [reply integrated_channels/blackboard/client.py:190]: Besides a 404 that is.

File: integrated_channels/blackboard/client.py
Comment (alex-sheehan-edx) [inline integrated_channels/blackboard/client.py:261]: I have mixed feelings about creating static methods to verify variable existence vs just checking inline, but I do feel like we should be consistent.

Comment (alex-sheehan-edx) [reply integrated_channels/blackboard/client.py:261]: How do you feel about either doing something like https://github.com/edx/edx-enterprise/pull/996/files#diff-e8224b0f5aaaee88283c12d5d533f5700f33f2e63c7939e93f814ef94bf95716R163? Or if not, then I don't see why we should have those static methods.

**Review by alex-sheehan-edx — COMMENTED** [review]
A couple of questions and comments but should be good afterwards! 👍

File: integrated_channels/blackboard/client.py
Comment (binodpant) [reply integrated_channels/blackboard/client.py:261]: yeah looks like that one is either way for me, it's a style thing. they are more accurately validator methods I think. stuff like does externalId exist? does course_id valid? and since they are used in more than once place I figured it would be better to DRY it up

I thikn if we called these validate_xyz instead of raise_for_ then it starts to make a bit more sense, what'cha think?

Comment (binodpant) [reply integrated_channels/blackboard/client.py:261]: the reason I think a method makes sense btw, is that it's very possible we want to validate more things about the payload in the future, so just thought would be a nice convenient place to put them (also makes it easier for us to tune the exception message in the future or do logging in one place, if we want to do those things for those cases)

File: integrated_channels/blackboard/client.py
Comment (binodpant) [reply integrated_channels/blackboard/client.py:190]: good point, more precisely really we could even just get the first item that is returned and use it without checking. I actually need to test this manually once, I __think__ it will work but I really have to be sure.. be back in a bit with this answer

File: integrated_channels/blackboard/client.py
Comment (binodpant) [reply integrated_channels/blackboard/client.py:56]: yes thank you!

File: integrated_channels/blackboard/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/blackboard/client.py:189]: My edge case questions would be:
- what's the response with no courses present (will we even have to worry about 404's with the error handling?)
- what's the  response with other courses but not the one we're looking for (will we even have to worry about 404's with the error handling?)
- what's the successful response body when we find just our course and just our course exists (I've tested this, it's not really an edge case)
- what's the successful response body when there are multiple courses, including our's?

If those are consistent, then I think we could just go without checking!

**Review by alex-sheehan-edx — APPROVED** [review]
Last bit of response on the comments, but otherwise LGTM once tests pass!

File: integrated_channels/blackboard/client.py
Comment (binodpant) [reply integrated_channels/blackboard/client.py:187]: yeah I would think that for if any reason we can't find a course_id here we can bail out safely

Comment (binodpant) [reply integrated_channels/blackboard/client.py:187]: and tested this code just now works!

**Pull request merged**
